### ISSUE TITLE
MAPE Memory Leak

### ISSUE THREAD

**Issue by jhkonan** [Lightning-AI/metrics#1160]
The `MAPE = torchmetrics.MeanAbsolutePercentageError().cuda()` has a memory leak.

This is my first time using torchmetrics. After much debugging, I have isolated MAPE as the source of the problem. The only reason this loop is written so explicitly is to isolate exactly where the issue is.

```py
sL1_list       = []
mae_list       = []
mse_list       = []
mape_list      = []

for e in tqdm(range(100)):
    
    torch.cuda.empty_cache()
    gc.collect()
    
    SL1  = torch.nn.SmoothL1Loss()
    MAE  = torch.nn.L1Loss()
    MSE  = torch.nn.MSELoss()
    MAPE = torchmetrics.MeanAbsolutePercentageError().cuda()
    
    loss           = None
    logits         = None
    spec           = None
    lld            = None
    temp_sL1_list  = []
    temp_mae_list  = []
    temp_mse_list  = []
    temp_mape_list = []
    
    #### TRAIN
    model.train()
    
    for i, (spec, lld, _) in enumerate(tqdm(train_dataloader)):
        
        spec, lld = spec.cuda(), lld.cuda()

        optimizer.zero_grad()
        logits = model(spec)[0][:, :396]
        loss = SL1(logits, lld)
        loss.backward()
        optimizer.step()
        
        break
    
    #### VALID
    model.eval()
        
    for i, (spec, lld, _) in enumerate(tqdm(train_dataloader)):
        
        spec, lld = spec.cuda(), lld.cuda()

        logits = model(spec)[0][:, :396]
        
        temp_sL1_list.append(SL1(logits, lld).item())
        temp_mae_list.append(MAE(logits, lld).item())
        temp_mse_list.append(MSE(logits, lld).item())
#         temp_mape_list.append(MAPE(logits, lld).item())
        
        break
```

Commenting out the one line for MAPE makes the memory leak go away. 

If you look at gc.get_objects(), there is a ton of garbage made by MAPE that I don't know how to get rid of.

Moving gc.collect() after the MAPE definition seems to fix the issue. We should not be required to re-instantiate the loss every epoch and garbage collect immediately after it is defined.

**Comment by justusschock** [comment]
Hi, Thanks for the issue.

A few things:
1. You are right, you don't have to re-instantiate it every epoch
2. You should call `.reset()` at the end of each epoch though
3. You do not calculate it with a `with torch.no_grad()` guard and you don't detach the results either. So maybe the memory leak is simply coming from the computational graph?

**Comment by jhkonan** [comment]
In this case, MAPE is used when the model is in eval mode, so no further grad accounting should be necessary.

Re-instantiating is required if gc.collect() is used, but I will try .reset() instead.

If reset is required for MAPE, please include it in the documentation or the example usage.

Thank you for your help!

**Comment by justusschock** [comment]
Eval-Mode of the model does not prevent it from gradient calculation, it only changes the way certain layers work. You probably still want to wrap a `with torch.no_grad` around it.

The `reset` is required for every metric, as you can see in our quickstart tutorial [here](https://torchmetrics.readthedocs.io/en/stable/pages/quickstart.html#module-metrics)

**Comment by jhkonan** [comment]
Thank you for this explanation. Everything is much clearer now — you do great work!
